Revert making a copy of RefCore in RefVector calls

Dr15Jones opened the pull request:
The clang check change made previously incorrectly converted a needed copy to just be a const ref. This ultimately lead to a threading race-condition when that code was called.

Dr15Jones committed:
Revert making a copy of RefCore in RefVector calls

The clang check change made previously incorrectly converted a needed copy to just be a const ref. This ultimately lead to a threading race-condition when that code was called.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-21772/2736

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

DataFormats/Common


@cmsbuild can you please review it and eventually sign? Thanks.
@makortel, @ebrondol, @rovere, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
once it gets through the jenkins tests, please also submit a backport to 94X.

Dr15Jones commented:
I was already making the backport.

cmsbuild commented:
+1
Tested at: 36e913350375c5d8e46c2cda786f4ff8fed52742
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21772/25255/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/9a782422a94819ff938c05b8e9370613c5473440
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21772/25255/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21772/25255/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21772/25255/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21772/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2773742
 - DQMHistoTests: Total failures: 1765
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2771808
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.05000000006 KiB( 22 files compared)
 - Checked 110 log files, 9 edm output root files, 26 DQM output files

davidlange6 commented on DataFormats/Common/interface/RefVector.h:
```diff
@@ -77,7 +77,7 @@ namespace edm {
       void const* memberPointer = refVector_.cachedMemberPointer(idx);
       if(memberPointer) {
-        const RefCore& newCore(core);
+        RefCore newCore(core);
```
hi @Dr15Jones - please add // NOLINT to each line so that clang-tidy doesn't revert these back at some point (here, its not really running as you just changed a header file)

dan131riley replied:
Hi @Dr15Jones @davidlange6 

Late comment: I think it would be clearer to make a non-const version of `RefCore::setProductPtr()` (with a different name) for use here.  That would let the compiler and any other tools know that the copy can't be optimized away, communicating intent in a way that `// NOLINT` doesn't.

Dr15Jones committed:
Stop clang-tidy from making change

clang-tidy incorrectly wants to change the copy into a const &. Adding the comment will prevent clang-tidy from affecting those lines.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-21772/2745

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

Dr15Jones commented:
@davidlange6 done

cmsbuild commented:
Pull request #21772 was updated. @cmsbuild can you please check and sign again.

davidlange6 commented:
merge

cmsbuild commented:
+1
Tested at: 29eb96ab7872bf1ce5f257f2d4a2ab98e8248007
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21772/25263/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21772/25263/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21772/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2773742
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2773572
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.12999999996 KiB( 22 files compared)
 - Checked 110 log files, 9 edm output root files, 26 DQM output files
